### PR TITLE
Reduce the indentation of (polytype) type constraints

### LINKED ISSUE #2143: Indentation feels off for mutually recursive functions with type constraints
The indentation feels off in the following example:
```ocaml
let rec equal_list :
          'a. ('a, 't) gexpr marked list -> ('a, 't) gexpr marked list -> bool =
 fun es1 es2 ->
  try List.for_all2 equal es1 es2 with Invalid_argument _ -> false

and equal : 'a. ('a, 't) gexpr marked -> ('a, 't) gexpr marked -> bool =
  fun (type a) (e1 : (a, 't) gexpr marked) (e2 : (a, 't) gexpr marked) ->
   match Marked.unmark e1, Marked.unmark e2 with
   …
```
* indentation for the type of `equal_list` is huge
* first `fun` is indented 1 and its body 2, it feels weird to have width-1 increments. Why not 2 and 4, or, better, 2 and 2 ? I'd even be ok with 0 and 2
* the second `fun` is indented 2 and its body 3, which is one more than for the first function and doesn't feel consistent

```editor-config
profile = default
margin = 80
exp-grouping = preserve
break-fun-decl = fit-or-vertical
wrap-comments
parse-docstrings
version=0.24.1
cases-exp-indent=2
indicate-multiline-delimiters=no
parens-tuple=multi-line-only
space-around-lists=false
break-infix-before-func
break-infix= fit-or-vertical
```

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #2143

### CODE REVIEW

File: lib/Fmt_ast.ml [hunk 1]
```diff
@@ -4320,7 +4320,7 @@ and fmt_value_binding c ~rec_flag ?ext ?in_ ?epi
   in
   let cstr_indent =
     match lb_typ with
-    | `Other {ast= {ptyp_desc= Ptyp_poly _; _}; _} -> 6
+    | `Other {ast= {ptyp_desc= Ptyp_poly _; _}; _} -> 3
```
Comment (Julow) [inline lib/Fmt_ast.ml:4323]: I think this should be 4. The indentation of 1 compared to the function body feels weird. Also, that's the indentation of other kinds of type annotations, which would remove the special case entirely.

File: CHANGES.md [hunk 2]
```diff
@@ -25,6 +25,7 @@ Tags:
 - \* Removed extra break in constructor declaration with comment (#2429, @Julow)
 - \* De-indent the `object` keyword in class types (#2425, @Julow)
 - \* Consistent formatting of arrows in class types (#2422, @Julow)
+- Reduce the indentation of (polytype) type constraints (#2437, @gpetiot)
```
Comment (Julow) [inline CHANGES.md:28]: It's a large change
```suggestion
- \* Reduce the indentation of (polytype) type constraints (#2437, @gpetiot)
```

**Review by Julow — COMMENTED** [review]
I'm not against this change but this causes huge diffs, I think we can take the time to investigate different solutions.

What do you think of this ?
```ocaml
  let t_of_sexp : 'a 'perms.
      (sexp -> 'a) -> (sexp -> 'perms) -> sexp -> ('a, 'perms) t =
```

File: lib/Fmt_ast.ml (hunk 1 above)
Comment (gpetiot) [reply lib/Fmt_ast.ml]: I changed the indent to 4 in the last commit to have a look, looks good to me as well, but 6 was a bit excessive.

Comment (Julow) [reply lib/Fmt_ast.ml]: The match and the variable can be removed entirely now.

**Comment by gpetiot** [comment]
> I'm not against this change but this causes huge diffs, I think we can take the time to investigate different solutions.
> 
> What do you think of this ?
> 
> ```ocaml
>   let t_of_sexp : 'a 'perms.
>       (sexp -> 'a) -> (sexp -> 'perms) -> sexp -> ('a, 'perms) t =
> ```

I tried this solution as well (see my last commit), I like it, but doesn't Jane Street want to have the `:` at the beginning of the line in this case?

**Comment by Julow** [comment]
I like the new change :) It's a good time to make it has the formatting of polytypes is changing anyway. I wonder what other think ?

> but doesn't Jane Street want to have the : at the beginning of the line in this case?

Yes, I believe this is controlled by the `break_colon` option and we must not break that.

File: test/passing/tests/polytypes-default.ml.ref [hunk 3]
```diff
@@ -1,25 +1,25 @@
 let t1 : 'a 'b. 'a t -> b t = ()
 
 let t2 :
-      'a 'b.
-      'a t________________________________ ->
-      'b t_______________________________________ =
+    'a 'b.
```
Comment (Julow) [inline test/passing/tests/polytypes-default.ml.ref:4]: The formatting between locally abstract types and `Ptyp_poly` is different:
```
let foo : type a b c.
    a -> b -> c foooooooo
    =
  ()

let foo :
    'a 'b 'c.
    a -> b -> c foooooooo
    =
  ()
```

I don't know whether this is important or not. The former is really part of the let-binding while the later is part of the type expression.

**Review by Julow — APPROVED** [review]
What do others think ?

File: test/passing/tests/polytypes-default.ml.ref (hunk 3 above)
Comment (gpetiot) [reply test/passing/tests/polytypes-default.ml.ref:4]: I forgot how the whole thing was designed, one day let's make the time to rewrite these in the parser directly so we can remove the abomination code in Sugar.ml

Comment (gpetiot) [reply test/passing/tests/polytypes-default.ml.ref:4]: Do you agree in merging it as-is? we can still improve the other case in a separate PR, or better redesign the types later.

**Pull request merged**